jpa/subsystem/src/main/java/org/jboss/as/jpa/subsystem/JPADeploymentDefinition.java

```diff
@@ -0,0 +1,40 @@
+package org.jboss.as.jpa.subsystem;
+
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.PathElement;
+import org.jboss.as.controller.SimpleResourceDefinition;
+import org.jboss.as.controller.descriptions.ModelDescriptionConstants;
+import org.jboss.as.controller.registry.ManagementResourceRegistration;
+import org.jboss.dmr.ModelNode;
+
+import static org.jboss.as.jpa.subsystem.JPADefinition.DEFAULT_DATASOURCE;
+import static org.jboss.as.jpa.subsystem.JPADefinition.DEFAULT_EXTENDEDPERSISTENCE_INHERITANCE;
+
+public class JPADeploymentDefinition extends SimpleResourceDefinition {
+
+    JPADeploymentDefinition() {
+        super(getParameters());
+    }
+
+    private static Parameters getParameters() {
+        Parameters result = new Parameters(PathElement.pathElement(ModelDescriptionConstants.SUBSYSTEM, JPAExtension.SUBSYSTEM_NAME),
+                JPAExtension.getResourceDescriptionResolver())
+                .setFeature(false)
+                .setRuntime();
+        return result;
+    }
+
+    @Override
+    public void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+        resourceRegistration.registerReadOnlyAttribute(DEFAULT_DATASOURCE, (context, operation) -> {
+            PathAddress subsystemAddress = PathAddress.pathAddress(PathElement.pathElement(ModelDescriptionConstants.SUBSYSTEM, JPAExtension.SUBSYSTEM_NAME));
+            ModelNode modelNode = context.readResourceFromRoot(subsystemAddress, false).getModel().get(DEFAULT_DATASOURCE.getName());
+            context.getResult().set(modelNode);
+        });
+        resourceRegistration.registerReadOnlyAttribute(DEFAULT_EXTENDEDPERSISTENCE_INHERITANCE, (context, operation) -> {
+            PathAddress subsystemAddress = PathAddress.pathAddress(PathElement.pathElement(ModelDescriptionConstants.SUBSYSTEM, JPAExtension.SUBSYSTEM_NAME));
+            ModelNode modelNode = context.readResourceFromRoot(subsystemAddress, false).getModel().get(DEFAULT_EXTENDEDPERSISTENCE_INHERITANCE.getName());
+            context.getResult().set(modelNode);
+        });
```

Comment:
The logic in these operation handlers will not work in domain mode, since the subsystem resources are not found at the root.
Rather, I would have these handlers delegate to the actual read-attribute operation handler as registered with the subsystem resource, so that they correctly handle operation flags, like expression resolution, etc.

Reply:
Can you perhaps explain a bit what needs to be done to tackle this request @pferraro ?

I've been trying to change the code the past week in order to "delegate to the actual read-attribute operation handler as registered with the subsystem resource" and I'm getting either `StackOverflow`, `IllegalArgumentException` or any other exception no matter what I do.

However, if I keep these changes, then it works just fine even in domain mode. For example, when I change the two attributes to non-default value ("ExampleDS" and "SHALLOW") and read the JPA subsystem in `/host=primary/server=server-one/subsystem=jpa:read-resource`, I'm getting

```
"outcome" => "success",
    "result" => {
        "default-datasource" => "ExampleDS",
        "default-extended-persistence-inheritance" => "SHALLOW"
```

The same outcome when I read the attribute in the deployment `/host=primary/server=server-one/deployment=kitchensink.war/subsystem=jpa:read-resource`. When I change the `default-datasource` to an expression `${whatever.test:ExampleDS}` then after resolving attributes in BOTH the deployment and the subsystem, I'm getting either "ExampleDS" or "NonExampleDS" (if the `whatever.test` is indeed "NonExampleDS").

From my point of view, everything works as intended. What am I missing?

Reply:
```java
        for (AttributeDefinition attribute : List.of(DEFAULT_DATASOURCE, DEFAULT_EXTENDEDPERSISTENCE_INHERITANCE)) {
            resourceRegistration.registerReadOnlyAttribute(attribute, new OperationStepHandler() {
                @Override
                public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
                    PathAddress subsystemAddress = context.getCurrentAddress().getParent().getParent().append(ADDRESS);
                    context.addStep(Util.getReadAttributeOperation(subsystemAddress, attribute.getName()), ReadAttributeHandler.INSTANCE, context.getCurrentStage(), true);
                }
            });
        }
```

Reply:
Thank you, I'll test it manually to see if it behaves the same in standalone as well as in domain